hack/test-go.sh

```diff
@@ -136,56 +144,51 @@ if [[ $iterations -gt 1 ]]; then
-    cover_profile=${cover_report_dir}/${arg}/coverage.out
-    mkdir -p "${cover_report_dir}/${arg}"
-    cover_params=(-cover -covermode="${KUBE_COVERMODE}" -coverprofile="${cover_profile}")
-  fi
-
+# If we're not collecting coverage, run all requested tests with one 'go test'
+# command, which is much faster.
+if [[ -z "${KUBE_COVER}" ]]; then
+  kube::log::status "Running unit tests without code coverage"
   go test "${goflags[@]:+${goflags[@]}}" \
-      ${KUBE_RACE} \
-      ${KUBE_TIMEOUT} \
-      "${cover_params[@]+${cover_params[@]}}" \
-      "${pkg}"
-  if [[ -f "${cover_profile}" ]]; then
-    # Include all coverage reach data in the combined profile, but exclude the
-    # 'mode' lines, as there should be only one.
-    grep -h -v "^mode:" ${cover_profile} >>${combined_cover_profile} || true
-  fi
-done
+    ${KUBE_RACE} ${KUBE_TIMEOUT} "${@+${@/#/${KUBE_GO_PACKAGE}/}}"
+  exit 0
+fi
 
-if [[ -f ${combined_cover_profile} ]]; then
-  coverage_html_file="${cover_report_dir}/combined-coverage.html"
-  go tool cover -html="${combined_cover_profile}" -o="${coverage_html_file}"
-  kube::log::status "Combined coverage report: ${coverage_html_file}"
-  if [[ -x "${KUBE_GOVERALLS_BIN}" ]]; then
-    ${KUBE_GOVERALLS_BIN} -coverprofile="${combined_cover_profile}" || true
-  fi
+# Create coverage report directories.
+cover_report_dir="/tmp/k8s_coverage/$(kube::util::sortable_date)"
+cover_profile="coverage.out"  # Name for each individual coverage profile
+kube::log::status "Saving coverage output in '${cover_report_dir}'"
+mkdir -p "${@+${@/#/${cover_report_dir}/}}"
+
+# Run all specified tests, collecting coverage results. Go currently doesn't
+# support collecting coverage across multiple packages at once, so we must issue
+# separate 'go test' commands for each package and then combine at the end.
+# To speed things up considerably, we can at least use xargs -P to run multiple
+# 'go test' commands at once.
+printf "%s\n" "${@}" | xargs -I{} -n1 -P${KUBE_COVERPROCS} \
+    go test "${goflags[@]:+${goflags[@]}}" \
+        ${KUBE_RACE} \
+        ${KUBE_TIMEOUT} \
+        -cover -covermode="${KUBE_COVERMODE}" \
+        -coverprofile="${cover_report_dir}/{}/${cover_profile}" \
+        "${cover_params[@]+${cover_params[@]}}" \
+        "${KUBE_GO_PACKAGE}/{}"
+
+combined_cover_profile="${cover_report_dir}/combined-coverage.out"
+# The combined coverage profile needs to start with a line indicating which
+# coverage mode was used (set, count, or atomic). This line is included in
+# each of the coverage profiles generated when running 'go test -cover', but
+# we strip these lines out when combining so that there's only one.
+echo "mode: ${KUBE_COVERMODE}" >"${combined_cover_profile}"
+
+# Include all coverage reach data in the combined profile, but exclude the
+# 'mode' lines, as there should be only one.
+find "${cover_report_dir}" -name "${cover_profile}" | \
+  xargs -I{} grep -h -v "^mode:" {} >>"${combined_cover_profile}" || true
```

Comment:
Use a single xargs grep, no reason to do file by file.

Also, a single redirect:

```shell
combined_cover_profile="${cover_report_dir}/combined-coverage.out"
{
  echo "mode: ${KUBE_COVERMODE}"
  find "${cover_report_dir}" -name "${cover_profile}" | xargs -r grep -h -v "^mode:"
} >"${combined_cover_profile}"
```